### ISSUE TITLE
wesnoth.sync.evaluate_single

### ISSUE THREAD

**Issue by sevu** [wesnoth/wesnoth#8960]
wesnoth.sync.evaluate_single behaves differently in 1.18 than in 1.16

In 1.16 I used the code below to check in an OOS-safe way weather an side is AI-controlled.
```
[if]
    [lua]
        code = <<
                return wesnoth.sync.evaluate_single(
                    function()
                        return {value = false}
                    end,
                    function()
                        return {value = true}
                    end
                ).value
        >>
    [/lua]
    [then]
       …
```

In 1.18 it can cause OOS by allowing undoing for one side, but not the other – independent of the result of the function, when the [then] tag was not even entered.

To reproduce, play a map with 2 sides in a networked game, e.g. Dark Forecast where they are allied.
1) Move a unit 2 fields – immediately synced to other player because of clearing fog
2) Move one field back – also synced to other player in 1.18 (not in 1.16)
3) Undo the move. That would be fine, because in 2) no new information was revealed. Move is undone for first player, but not for 2nd.

So, if this Lua code is in a moveto event it makes every move undoable, because checking the condition is now a gamestate change. This is in stable. cannot be changed.

Is there another way to check OOS-safe if a side has an ai controller? My best idea is currently to use the code from above in a turn start event, temporarily save the result in a variable and check the variable in moveto events.

**Comment by Pentarctagon** [comment]
@CelticMinstrel would this potentially be related to the event handling overhaul?

**Comment by gfgtdf** [comment]
> To reproduce, play a map with 2 sides in a networked game, e.g. Dark Forecast where they are allied.
> 
>     1. Move a unit 2 fields – immediately synced to other player because of clearing fog
> 
>     2. Move one field back – also synced to other player in 1.18 (not in 1.16)
> 
>     3. Undo the move. That would be fine, because in 2) no new information was revealed. Move is undone for first player, but not for 2nd.

This described beahvior is buggy in multilpe levels

1. `sync.evaluate_single` shouldn't casue an action to be sent
2. if we do send an action the engine should automaticially block undoing regardless of a possible [allow_undo]


> So, if this Lua code is in a moveto event it makes every move undoable, because checking the condition is now a gamestate change. This is in stable. cannot be changed.

Imo If the current behavior causes OOS anyways we should change it. (even more if it's a regression from 1.16)

**Comment by gfgtdf** [comment]
also, could you please test wehether the bug is also in 1.19/master ?

**Comment by gfgtdf** [comment]
in particular problem (2) is strange since the code that sends the data, explicitly checks whether the undo stack is clear before attemoting to send things like movement :

see https://github.com/wesnoth/wesnoth/blob/1.18.0/src/playturn.cpp#L82

**Comment by CelticMinstrel** [comment]
> @CelticMinstrel would this potentially be related to the event handling overhaul?

Sounds like it maybe could be related to changes surrounding the undo API.

**Comment by sevu** [comment]
Affects 1.19 too.. I wrote [if] above, but what I used is [filter_condition].

Put this into Dark Forcast:
```
[event]
    name=moveto
    first_time_only=no
    [filter_condition]
        [lua]
            code = <<
                    return wesnoth.sync.evaluate_single(
                        function()
                            return {value = false}
                        end,
                        function()
                            return {value = false}
                        end
                    ).value
            >>
        [/lua]
    [/filter_condition]

    [message]
        speaker=narrator
        message= _ "Unreachable code."
    [/message]
[/event]
```

**Comment by ProditorMagnus** [comment]
Why it needs dark forecast?

**Comment by sevu** [comment]
Doesn't. But good for testing because 2 allied sides.

**Comment by ProditorMagnus** [comment]
What it is supposed to test? Code review says message is unreachable, and message text agrees, and game doesnt show this message.

**Comment by sevu** [comment]
> Move a unit 2 fields – immediately synced to other player because of clearing fog
> Move one field back – also synced to other player in 1.18 (not in 1.16)
> Undo the move. That would be fine, because in 2) no new information was revealed. Move is undone for first player, but not for 2nd.

This leads to OOS if you move in the next turn a unit onto the place, where it is free for you, but occupied in the other client.

**Comment by ProditorMagnus** [comment]
For the undo example you had one branch returning true.

**Comment by sevu** [comment]
Yes, While testing I used always the first branch though (me being human and doing the moves, not 2nd).

Also never tested case with [if], was a mistake to not write [filter_condition] in original post. I think that might be relevant to the link by gfgtdf.

**Comment by gfgtdf** [comment]
> I think that might be relevant to the link by gfgtdf.

It actually shouldn't. what the code does(or apparently  doesn't) is, it makes sure no actions are sent to other clients when the undo stack is nonempty preventing these kinds of oos. And this should be completely independent on what the event related (wml, Lua or c++) code does.

In 1.19 this code is even simpler (that's why I asked whether it also effects 1.19) making this even more strange.

**Comment by gfgtdf** [comment]
This is the code that makes sure the undo stack is cleared in case of undo_blocked() 

https://github.com/wesnoth/wesnoth/blob/971073055e2487560940790cc340d91cd8d8e8df/src/actions/move.cpp#L1162

**Comment by sevu** [comment]
There is another change in 1.18 which reminds me of this case, though that one I consider a feature:

[message][option] displays now a message to other players while choosing. If the chosen option was a do-nothing option with [allow_undo], then undoing the move won't be undone on other clients. (Presumably because the waiting message synced).

Could this be related to this case?

**Comment by gfgtdf** [comment]
yes, seems like the same issue to me :/

**Comment by gfgtdf** [comment]
> though that one I consider a feature:

It also cases oos no?

**Comment by gfgtdf** [comment]
Hmm okay, i think i have an idea whats happening:  this code 

https://github.com/wesnoth/wesnoth/blob/971073055e2487560940790cc340d91cd8d8e8df/src/synced_context.cpp#L215 

returns the wrong result, because (randomness::generator->get_random_calls() != 0) sometimes uses the unsynced rng (which always has non zero random calls) (i cannot build wesnot here so i cannot really confirm this theory but it makes a lot of sense to me)

**Comment by gfgtdf** [comment]
Can you test wheterh https://github.com/wesnoth/wesnoth/pull/8983 fixes the issue for you ?

**Comment by gfgtdf** [comment]
(the patch can also be applied to the 1.18 it shoudl be the same)

**Comment by sevu** [comment]
Tested in 1.18:
- wesnoth.sync.evaluate_single fixed
- [message][option][allow_undo] fixed
- [message]wait_description is not displayed with the patch

Not sure if you can have both [allow_undo] being possible and having wait_description.

**Comment by gfgtdf** [comment]
Okay i made a pr for 1.18 https://github.com/wesnoth/wesnoth/pull/8985 which i intend to merge as soon as the CI is done.

For 1.19 i'm probabyl just removing that line instead.

**Comment by gfgtdf** [comment]
Hmm i cannot login into discord right now, but it seems like 1.18.1 was tagged right before this fix which is a bit unfortunate

**Comment by CelticMinstrel** [comment]
#8983 forward-ported a fix for this, but this issue is being left open because @gfgtdf thought there may be a better way to handle it.

**Comment by Wedge009** [comment]
Should this still be left open? Not sure if @gfgtdf is working on the 'better way'.
